src/hippopt/base/optimization_problem.py

```diff
@@ -0,0 +1,103 @@
+import abc
+import dataclasses
+import types
+from enum import Enum
+from typing import Generator, List
+
+import casadi as cs
+
+from hippopt.base.opti_solver import OptiSolver
+from hippopt.base.optimization_object import OptimizationObject, TOptimizationObject
+from hippopt.base.optimization_solver import OptimizationSolver, TOptimizationSolver
+
+
+class ExpressionType(Enum):
+    skip = 0
+    subject_to = 1
+    minimize = 2
+
+
+@dataclasses.dataclass
+class OptimizationProblem(abc.ABC):
+    optimization_solver: dataclasses.InitVar[OptimizationSolver] = dataclasses.field(
+        default=None
+    )
+    _solver: TOptimizationSolver = dataclasses.field(default=None)
+
+    def __post_init__(self, optimization_solver: TOptimizationSolver = None):
+        self._solver = (
+            optimization_solver
+            if isinstance(optimization_solver, OptimizationSolver)
+            else OptiSolver()
+        )
+
+    def generate_optimization_objects(
+        self, input_structure: OptimizationObject | List[TOptimizationObject]
+    ) -> TOptimizationObject | List[TOptimizationObject]:
+        return self._solver.generate_optimization_objects(
+            input_structure=input_structure
+        )
+
+    def add_cost(
+        self,
+        expression: cs.MX | Generator[cs.MX, None, None],
+        scaling: float | cs.MX = 1.0,
+    ):
```

Comment:
The problem would be that it is difficult to associate an entry of the list with a meaningful tag. In other words, if I have a method returning ``costs = list[cs.MX]``, how do I understand what is ``cost[5]``? For the time being, I leave this on the user.

Reply:
Also, this would complexify the rebase of a branch I have been working on. I have spun off an issue https://github.com/ami-iit/hippopt/issues/6 and I will be dealing with this on the next PR